packages/unlock-js/src/walletService.ts

```diff
@@ -997,4 +1013,150 @@ export default class WalletService extends UnlockService {
       callback
     )
   }
+
+  async swapAndCall({
+    params: {
+      network,
+      lockAddress,
+      srcTokenAddress,
+      amountInMax,
+      uniswapRouter,
+      callData,
+      swapCallData,
+    },
+  }: {
+    params: {
+      lockAddress: string
+      network: number
+      srcTokenAddress: string
+      amountInMax: string
+      uniswapRouter: string
+      swapCallData: string
+      callData: string
+    }
+  }) {
+    const networkConfig = this.networks[network]
+    const UnlockSwapPurchaser = networkConfig?.swapPurchaser
+    if (!UnlockSwapPurchaser) {
+      throw new Error('SwapPurchaser not available for this network')
+    }
+    const provider = this.providerForNetwork(network)
+    const swapPurchaserContract = new ethers.Contract(
+      UnlockSwapPurchaser,
+      UnlockSwapPurchaserABI,
+      provider
+    )
+    const swapPurchaser = swapPurchaserContract.connect(this.signer)
+
+    const tx = await swapPurchaser.swapAndCall(
+      lockAddress,
+      srcTokenAddress,
+      amountInMax,
+      uniswapRouter,
+      swapCallData,
+      callData,
+      {
+        value: amountInMax,
+      }
+    )
+    return tx
+  }
+
+  async swapAndPurchase({
+    params: {
+      lockAddress,
+      keyManagers: _keyManagers,
+      keyPrices: _keyPrices,
+      referrers: _referrers,
+      owners,
+      data: _data,
+      srcTokenAddress,
+      amountInMax,
+      uniswapRouter,
+      swapCallData,
+      network,
+    },
+  }: {
+    params: PurchaseKeysParams & {
+      srcTokenAddress: string
+      amountInMax: string
+      uniswapRouter: string
+      swapCallData: string
+      network: number
+    }
+  }) {
+    const lockContract = await this.getLockContract(lockAddress)
+    const erc20Address = await lockContract.tokenAddress()
+    const provider = this.providerForNetwork(network)
+    const networkConfig = this.networks[network]
+    const swapPurchaser = networkConfig.swapPurchaser!
+
+    const decimals =
+      erc20Address !== ethers.constants.AddressZero
+        ? await getErc20Decimals(erc20Address, provider)
+        : networkConfig.nativeCurrency.decimals
+    // we parse by default a length corresponding to the owners length
+    const defaultArray = Array(owners.length).fill(null)
+
+    const keyPrices = await Promise.all(
+      (_keyPrices || defaultArray).map(async (kp) => {
+        if (!kp) {
+          // We might not have the keyPrice, in which case, we need to retrieve from the the lock!
+          return await lockContract.keyPrice()
+        }
+        return formatKeyPrice(kp, erc20Address, decimals, provider)
+      })
+    )
+
+    const keyManagers = (_keyManagers || defaultArray).map(
+      (km) => km || ethers.constants.Zero
+    )
+
+    const referrers = (_referrers || defaultArray).map(
+      (km) => km || ethers.constants.Zero
+    )
+
+    const data = (_data || defaultArray).map((d) => d || [])
+
+    const args = [keyPrices, owners, referrers, keyManagers, data]
```

Comment:
This ^^ can all be taken from `purchase`, no? Since this is also version specific I think it would be really nice to just extract that from the existing purchase function in each version to `buildPurchaseArgs` and just use it here?